Add missing cs_main locks when accessing chainActive

practicalswift opened the pull request:
* Add missing `cs_main` locks when accessing `chainActive`. (The variable `chainActive` is guarded by the mutex `cs_main`.)
* Add corresponding annotations (`GUARDED_BY` + `EXCLUSIVE_LOCKS_REQUIRED`).

sdaftuar commented on src/validation.cpp:
```diff
@@ -2669,8 +2669,11 @@ static bool ReceivedBlockTransactions(const CBlock &block, CValidationState& sta
-            if (chainActive.Tip() == nullptr || !setBlockIndexCandidates.value_comp()(pindex, chainActive.Tip())) {
-                setBlockIndexCandidates.insert(pindex);
+            {
+                LOCK(cs_main);
```
I think we can just add an `AssertLockHeld(cs_main)` to the top of this function -- this function only has two callers, both of which lock cs_main before calling.

morcos commented on src/wallet/wallet.cpp:
```diff
@@ -2648,7 +2651,10 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
-    assert(txNew.nLockTime <= (unsigned int)chainActive.Height());
+    {
+        LOCK(cs_main);
+        assert(txNew.nLockTime <= (unsigned int)chainActive.Height());
```
I think we should just get rid of this assert.  We can't guarantee chainActive height will not go down.

practicalswift commented:
@sdaftuar @morcos Thanks for reviewing. Suggested changes incorporated! :-)

promag commented on src/wallet/wallet.cpp:
```diff
@@ -2639,7 +2639,10 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
     // nLockTime that preclude a fix later.
-    txNew.nLockTime = chainActive.Height();
+    {
+        LOCK(cs_main);
```
Remove, all callers have `cs_main` locked, assert lock is held instead. Ping @ryanofsky as he says the goal is to removed recursive locks.

practicalswift replied:
@promag Thanks! Now fixed.

practicalswift replied:
@promag I'm getting an assertion error when running with `test/functional/test_runner.py` with your suggested patch applied. Could you investigate? Reverting to previous version for now.

promag replied:
I'll check.

practicalswift replied:
@promag Any idea on how to tackle the assertion failure? :-)

practicalswift replied:
@promag See @TheBlueMatt's comment about `cs_main` locking in `CWallet::CreateTransaction(…)`: https://github.com/bitcoin/bitcoin/pull/11596#discussion_r149189367

promag reviewed (commented):
utACK.

promag commented:
@practicalswift following https://github.com/bitcoin/bitcoin/pull/11596#discussion_r149104246, I believe a7eb21a is the fix.

practicalswift commented:
@promag What is the reasoning behind the patch – what are the underlying locking requirements? Your patch is likely correct, but I'm trying to understand exactly why :-)

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -2594,6 +2594,8 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
 bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletTx& wtxNew, CReserveKey& reservekey, CAmount& nFeeRet,
                                 int& nChangePosInOut, std::string& strFailReason, const CCoinControl& coin_control, bool sign)
 {
+    AssertLockHeld(cs_main);
```
Ugh, can we not add more cs_main requirements - if the goal is to resolve the chainActive.Height() issue only then please add the cs_main around just that, its not gonna hurt anything to not hold cs_main for the duration of this function, and we really should be trying to *remove* cs_main from this function, not go the other way.

practicalswift commented:
@TheBlueMatt Thanks for reviewing. Feedback addressed.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3015,8 +3015,11 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
-    if (!pwallet->FundTransaction(tx, nFeeOut, changePosition, strFailReason, lockUnspents, setSubtractFeeFromOutputs, coinControl)) {
-        throw JSONRPCError(RPC_WALLET_ERROR, strFailReason);
+    {
+        LOCK2(cs_main, pwallet->cs_wallet);
```
Please no. Can we instead add the locking in CreateTransaction itself? We need to be pushing locks down, especially when it comes to cs_main, not putting more locks in RPC functions.

practicalswift commented:
Reverting the patch suggested by @promag. The locking is now made down in `CreateTransaction` as suggested by @TheBlueMatt.

TheBlueMatt commented:
utACK 2e441c91e7806e6ae781ae0b0aefea958079d15b

luke-jr commented on src/rest.cpp:
```diff
@@ -510,7 +510,11 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
-        ssGetUTXOResponse << chainActive.Height() << chainActive.Tip()->GetBlockHash() << bitmap << outs;
+        {
+            LOCK(cs_main);
+            ssGetUTXOResponse << chainActive.Height() << chainActive.Tip()->GetBlockHash();
```
This isn't sufficient I think. We don't want to allow `chainActive` to change from when we get the results, to when we identify the chain they're for. So extract this info above in the first lock...

promag replied:
Which first lock?

practicalswift replied:
@luke-jr Thanks for reviewing. Do you mean like this?

```patch
diff --git a/src/rest.cpp b/src/rest.cpp
index 0ed11e6..649059e 100644
--- a/src/rest.cpp
+++ b/src/rest.cpp
@@ -479,6 +479,8 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
     std::string bitmapStringRepresentation;
     std::vector<bool> hits;
     bitmap.resize((vOutPoints.size() + 7) / 8);
+    int chainActiveHeight;
+    uint256 chainActiveTipBlockHash;
     {
         LOCK2(cs_main, mempool.cs);

@@ -503,6 +505,9 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
             bitmapStringRepresentation.append(hit ? "1" : "0"); // form a binary string representation (human-readable for json output)
             bitmap[i / 8] |= ((uint8_t)hit) << (i % 8);
         }
+
+        chainActiveHeight = chainActive.Height();
+        chainActiveTipBlockHash = chainActive.Tip()->GetBlockHash();
     }

     switch (rf) {
@@ -510,11 +515,7 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
         // serialize data
         // use exact same output as mentioned in Bip64
         CDataStream ssGetUTXOResponse(SER_NETWORK, PROTOCOL_VERSION);
-        {
-            LOCK(cs_main);
-            ssGetUTXOResponse << chainActive.Height() << chainActive.Tip()->GetBlockHash();
-        }
-        ssGetUTXOResponse << bitmap << outs;
+        ssGetUTXOResponse << chainActiveHeight << chainActiveTipBlockHash << bitmap << outs;
         std::string ssGetUTXOResponseString = ssGetUTXOResponse.str();

         req->WriteHeader("Content-Type", "application/octet-stream");
@@ -524,11 +525,7 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)

     case RF_HEX: {
         CDataStream ssGetUTXOResponse(SER_NETWORK, PROTOCOL_VERSION);
-        {
-            LOCK(cs_main);
-            ssGetUTXOResponse << chainActive.Height() << chainActive.Tip()->GetBlockHash();
-        }
-        ssGetUTXOResponse << bitmap << outs;
+        ssGetUTXOResponse << chainActiveHeight << chainActiveTipBlockHash << bitmap << outs;
         std::string strHex = HexStr(ssGetUTXOResponse.begin(), ssGetUTXOResponse.end()) + "\n";

         req->WriteHeader("Content-Type", "text/plain");
@@ -541,11 +538,8 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)

         // pack in some essentials
         // use more or less the same output as mentioned in Bip64
-        {
-            LOCK(cs_main);
-            objGetUTXOResponse.push_back(Pair("chainHeight", chainActive.Height()));
-            objGetUTXOResponse.push_back(Pair("chaintipHash", chainActive.Tip()->GetBlockHash().GetHex()));
-        }
+        objGetUTXOResponse.push_back(Pair("chainHeight", chainActiveHeight));
+        objGetUTXOResponse.push_back(Pair("chaintipHash", chainActiveTipBlockHash.GetHex()));
         objGetUTXOResponse.push_back(Pair("bitmap", bitmapStringRepresentation));

         UniValue utxos(UniValue::VARR);
```

luke-jr commented on src/rest.cpp:
```diff
@@ -520,7 +524,11 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
-        ssGetUTXOResponse << chainActive.Height() << chainActive.Tip()->GetBlockHash() << bitmap << outs;
+        {
+            LOCK(cs_main);
+            ssGetUTXOResponse << chainActive.Height() << chainActive.Tip()->GetBlockHash();
```
Same

luke-jr commented on src/rest.cpp:
```diff
@@ -533,8 +541,11 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
+        {
+            LOCK(cs_main);
+            objGetUTXOResponse.push_back(Pair("chainHeight", chainActive.Height()));
+            objGetUTXOResponse.push_back(Pair("chaintipHash", chainActive.Tip()->GetBlockHash().GetHex()));
```
Same.

practicalswift commented:
Added commit 9b3d094894350933342b5e352785cfdd9f2d03fe per @luke-jr:s feedback.

Please re-review :-)

TheBlueMatt commented:
@practicalswift why did you rebase this? It makes reviewing something that was previously reviewed much more difficult. Mind squashing the two commits? There seems to be no reason to have them separate.

luke-jr reviewed (approved):
utACK

practicalswift commented:
Now squashed into one commit. Please re-review! :-)

@TheBlueMatt The reason for the rebase was that I wanted to get rebased on top of 76ea17c7964c15dd90e10c2c257cdeb5847b3d69 (IIRC) in order to get the build to pass – I always compile with `-Werror=thread-safety-analysis` :-)

The reason for the two commits was to allow separate review for the latter commit in the case that I had misunderstood @luke-jr:s suggestion and needed to revert.

practicalswift commented:
Updated:
* Added a few more missing `cs_main` locks.
* Added a commit with the Clang thread safety analysis annotations `chainActive GUARDED_BY(cs_main)` and the corresponding `EXCLUSIVE_LOCKS_REQUIRED(…)` annotations.

Please review :-)

TheBlueMatt commented:
The addition of annotations here isn't complete - do you not need to add a EXCLUSIVE_LOCKS_REQUIRED annotation to the declarations, not the definitions of functions?

On November 23, 2017 4:24:23 AM PST, practicalswift <notifications@github.com> wrote:
>Updated:
>* Added a few more missing `cs_main` locks.
>* Added a commit with the Clang thread safety analysis annotations
>`chainActive GUARDED_BY(cs_main)` and the corresponding
>`EXCLUSIVE_LOCKS_REQUIRED(…)` annotations.
>
>Please review :-)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/11596#issuecomment-346604992

practicalswift commented:
I've now pushed an updated version adding [a missing `OpenWallets()` lock](see https://github.com/bitcoin/bitcoin/pull/11596/files#diff-c865a8939105e6350a50af02766291b7R1575) and some additional annotations.

@TheBlueMatt - to make the annotations 100 % complete and moved to the declarations we'll have to handle some header files for which the existence of `cs_main` is currently not known. More specifically the following header files:
* `net_processing.h`
* `qt/transactiondesc.h`
* `qt/transactionrecord.h`
* `rpc/blockchain.h`
* `wallet/init.h`
* `wallet/wallet.h`

What is the most appropriate way to handle `cs_main` in respective cases above? I guess we don't want to sprinkle `extern CCriticalSection cs_main;` all over the place? :-)

TheBlueMatt commented:
@practicalswift No, we aren't adding the annotations just to add the annotations. In order for them to be really effective and provide the kind of guarantees we want they need to be listed everywhere the function which needs the annotation is declared, not just ate the definition. I'd rather we leave them out than add them just to function declarations. The annotations are going to be highly infective for things like cs_main - they're gonna propagate backwards through the callstack pretty deep, but that's ok, its how it should be. We shouldn't avoid it, just gotta take the dive (though preferably not in this PR, please, lets just get this one merged).

practicalswift commented:
@TheBlueMatt Is that an utACK for this PR as-is? :-)

TheBlueMatt commented:
If you remove the annotation commit to revert to where people had previously reviewed this, I'm happy to review again.

practicalswift commented:
@TheBlueMatt Reverted and pushed. Now only keeping the lock addiiton (removed adding annotations). Looks good? :-)

TheBlueMatt commented on src/init.cpp:
```diff
@@ -1581,8 +1582,11 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
-        return false;
+    {
+        LOCK(cs_main);
+        if (!OpenWallets())
```
I'd kinda prefer we push this lock down into wallet/init. Seemd weird to export the cs_main requirement to OpenWallets.

TheBlueMatt commented on src/init.cpp:
```diff
@@ -1492,6 +1492,7 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
                         strLoadError = _("Error initializing block database");
                         break;
                     }
+                    LOCK(cs_main);
```
There are a ton of cs_main-required things in this block, seems weird to put this all the way down here. For static-analysis purposes, its probably best to just take cs_main at the top of the try{} block and hold it through the whole chain-load process.

practicalswift commented:
@TheBlueMatt Thanks for reviewing! Feedback address. Would you mind re-reviewing? :-)

luke-jr commented on src/init.cpp:
```diff
@@ -1634,10 +1632,13 @@ bool AppInitMain()
 
     // Either install a handler to notify us when genesis activates, or set fHaveGenesis directly.
     // No locking, as this happens before any background thread is started.
```
This comment needs to be updated at least. Preferably with an explanation why you're adding a lock here...

luke-jr reviewed (approved):
utACK

practicalswift commented:
@luke-jr Good point! Comment updated. Please re-review :-)

practicalswift commented:
@MarcoFalke Good idea! Annotations added. Please review :-)

luke-jr commented:
Why did you un-remove the double-lock at "The block database contains a block which appears to be from the future."?

practicalswift commented:
@luke-jr I'm actually investigating a dead-lock right now which seems to have been introduced during the last rebase. Which double-lock are you referring to? Do you have a diff?

I'm investigating as we speak.

practicalswift commented:
@luke-jr You mean this one?

```
diff --git a/src/init.cpp b/src/init.cpp
index e7544e8..f18a947 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1429,7 +1429,6 @@ bool AppInitMain()
         nStart = GetTimeMillis();
         do {
             try {
-                LOCK(cs_main);
                 UnloadBlockIndex();
                 pcoinsTip.reset();
                 pcoinsdbview.reset();
```

luke-jr commented:
Search init.cpp for "`The block database contains a block which appears to be from the future.`". It grabs cs_main immediately before that. But cs_main is already held here now.

practicalswift committed:
Locking annotations: Variable chainActive is guarded by cs_main

practicalswift committed:
Locking annotations: Add missing cs_main locks (when accessing chainActive)

practicalswift committed:
Locking annotations: Add EXCLUSIVE_LOCKS_REQUIRED(...) annotations that follow from "chainActive GUARDED_BY(cs_main)"

practicalswift commented:
I've now split up this PR in three parts:
* Add missing `LOCK(cs_main)` where required for `chainActive`
* Annotation: Add `chainActive GUARDED_BY(cs_main)` – does **not** change run-time behaviour
* Annotation: Add `EXCLUSIVE_LOCKS_REQUIRED(...)` as implied by the `chainActive` guard  – does **not** change run-time behaviour

Please re-review :-)

/cc @sdaftuar @morcos @promag @TheBlueMatt @luke-jr @MarcoFalke

practicalswift committed:
Remove invalid assertion: We can't guarantee chainActive height will not go down

As noted by @marcos in https://github.com/bitcoin/bitcoin/pull/11596#discussion_r148614708

practicalswift commented:
Rebased!

practicalswift commented:
Rebased!

practicalswift commented:
Rebased! Please re-review :-)

TheBlueMatt commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -301,10 +312,16 @@ class ListCoinsTestingSetup : public TestChain100Setup
-        BOOST_CHECK(it != wallet->mapWallet.end());
-        it->second.SetMerkleBranch(chainActive.Tip(), 1);
+        std::map<uint256, CWalletTx>::iterator it;
+        {
```
Can you just take both locks instead of adding a new scope and lots of diff lines? Its in a test, locks shouldn't matter much.

TheBlueMatt commented on src/interfaces/wallet.cpp:
```diff
@@ -54,7 +54,7 @@ class PendingWalletTxImpl : public PendingWalletTx
 
 //! Construct wallet tx struct.
-WalletTx MakeWalletTx(CWallet& wallet, const CWalletTx& wtx)
+WalletTx MakeWalletTx(CWallet& wallet, const CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
a) annotations need to go in the header file, not the cpp file. b) I dont think we can put locks required in the interfaces file, as it may become a process boundary. The callee will have to take the lock itself.

ryanofsky replied:
> annotations need to go in the header file

This is ok, these are just internal functions (in an anonymous namespace) not shared or exposed in headers.

TheBlueMatt reviewed (commented):
I guess this is somewhat duplicative with #11652, no? My review comments there apply here too.

practicalswift commented:
Closing. Will add new PR based on #11652 and #13083.
